interfaces/builtin/wayland.go

```diff
@@ -19,32 +19,127 @@
 
 package builtin
 
+import (
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
 const waylandSummary = `allows access to compositors supporting wayland protocol`
 
 const waylandBaseDeclarationSlots = `
   wayland:
     allow-installation:
       slot-snap-type:
+        - app
         - core
+    deny-auto-connection:
+      on-classic: false
 `
 
-const waylandConnectedPlugAppArmor = `
-# Description: Can access compositors supporting the wayland protocol
+const waylandPermanentSlotAppArmor = `
+# Description: Allow operating as a Wayland display server. This gives privileged access
+# to the system.
+
+# needed since Wayland is a display server and needs to configure tty devices
+capability sys_tty_config,
+/dev/tty[0-9]* rw,
+
+# Create the Wayland socket and lock file
+/run/user/[0-9]*/wayland-[0-9]* rw,
+
+# Allow write access to create XDG_RUNTIME_DIR (until lp:1656340 is fixed)
+/run/user/[0-9]*/ w,
+
+# Needed for mode setting via drmSetMaster() and drmDropMaster()
+capability sys_admin,
+
+# Weston probes this on start
+/sys/devices/pci**/boot_vga r,
+
+# NOTE: this allows reading and inserting all input events
+/dev/input/* rw,
 
-# Allow access to the wayland compsitor server socket
+# For using udev
+network netlink raw,
+/run/udev/data/c13:[0-9]* r,
+/run/udev/data/+input:input[0-9]* r,
+/run/udev/data/+platform:* r,
+`
+
+const waylandPermanentSlotSecComp = `
+# Description: Allow operating as a Wayland server. This gives privileged access
+# to the system.
+# Needed for server launch
+bind
+listen
+# Needed by server upon client connect
+accept4
+# for udev
+socket AF_NETLINK - NETLINK_KOBJECT_UEVENT
+`
+
+const waylandConnectedPlugAppArmor = `
+# Allow access to the Wayland compositor server socket
 owner /run/user/*/wayland-[0-9]* rw,
 
+# XWayland needs access to this
+/run/xwayland-shared-* rw,
```

Comment:
You removed this from waylandPermanentSlotAppArmor, but left it here. This should be removed too, no?

Reply:
I'm building my Chromium snap to incorporate XWayland, which will create that socket on startup. I was thinking it handiest to add support for any XWayland client to the Wayland interface. But in retrospect that's not very secure.

I'll iterate this a bit more.

Reply:
Yes you right, this is pointless, removing.